src/tracing/esp32_diagnostic_trace/DiagnosticStorage.cpp

```diff
@@ -0,0 +1,106 @@
+/*
+ *
+ *    Copyright (c) 2025 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <tracing/esp32_diagnostic_trace/DiagnosticStorage.h>
+using namespace chip::TLV;
+
+namespace chip {
+namespace Tracing {
+namespace Diagnostics {
+
+CHIP_ERROR CircularDiagnosticBuffer::Store(const DiagnosticEntry & entry)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    mWriter.Init(*this);
+    ReturnLogErrorOnFailure(Encode(mWriter, entry));
+    return err;
+}
+
+CHIP_ERROR CircularDiagnosticBuffer::Retrieve(MutableByteSpan & span, uint32_t & read_entries)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    mReader.Init(*this);
+
+    TLVWriter writer;
+    writer.Init(span.data(), span.size());
+    read_entries = 0;
+
+    bool close_success                = true; // To check if the last TLV is copied successfully.
+    uint32_t successful_written_bytes = 0;    // Store temporary writer length in case last TLV is not copied successfully.
+
+    while ((err = mReader.Next()) == CHIP_NO_ERROR)
```

Comment:
do we need to handle CHIP_END_OF_TLV differently than others (i.e. end of TLV is essentially success read all, where as other documented errors like TLV_UNDERRUN or INVALID_TLV_ELEMENT are log-worthy failures)